Python/suggestions.c

```diff
@@ -7,74 +7,181 @@
     }
 
-    size_t b_index = 0;
     size_t result = 0;
-    while (b_index < b_size) {
+    for (size_t b_index = 0; b_index < b_size; b_index++) {
         char code = b[b_index];
-        size_t distance = result = b_index++;
-        index = SIZE_MAX;
-        while (++index < a_size) {
-            size_t b_distance = code == a[index] ? distance : distance + 1;
+        // cost(b[:b_index], a[:0]) == b_index * MOVE_COST
+        size_t distance = result = b_index * MOVE_COST;
+        size_t minimum = SIZE_MAX;
+        for (size_t index = 0; index < a_size; index++) {
+
+            // cost(b[:b_index+1], a[:index+1]) = min(
+            //     // 1) substitute
+            //     cost(b[:b_index], a[:index])
+            //         + substitution_cost(b[b_index], a[index]),
+            //     // 2) delete from b
+            //     cost(b[:b_index], a[:index+1]) + MOVE_COST,
+            //     // 3) delete from a
+            //     cost(b[:b_index+1], a[index]) + MOVE_COST
+            // )
+
+            // 1) Previous distance in this row is cost(b[:b_index], a[:index])
+            size_t substitute = distance + substitution_cost(code, a[index]);
+            // 2) cost(b[:b_index], a[:index+1]) from previous row
             distance = buffer[index];
-            if (distance > result) {
-                if (b_distance > result) {
-                    result = result + 1;
-                } else {
-                    result = b_distance;
-                }
-            } else {
-                if (b_distance > distance) {
-                    result = distance + 1;
-                } else {
-                    result = b_distance;
-                }
-            }
+            // 3) existing result is cost(b[:b_index+1], a[index])
+
+            size_t insert_delete = Py_MIN(result, distance) + MOVE_COST;
+            result = Py_MIN(insert_delete, substitute);
+
+            // cost(b[:b_index+1], a[:index+1])
             buffer[index] = result;
+            if (result < minimum) {
+                minimum = result;
+            }
+        }
+        if (minimum > max_cost) {
+            // Everything in this row is too big, so bail early.
+            return max_cost + 1;
         }
     }
-    PyMem_Free(buffer);
     return result;
 }
 
+#define TEST_LEVENSHTEIN 0
+#if TEST_LEVENSHTEIN
+
+static void
+do_test(const char *a, const char *b, size_t d)
+{
+    static size_t buffer[500];
+    size_t a_len = strlen(a);
+    size_t b_len = strlen(b);
+    size_t d2 = levenshtein_distance(a, a_len, b, b_len, buffer, 9999);
+    printf("'%s' -> '%s':", a, b);
+    if (d != d2) {
+        printf("\n    FAILURE: Expected %d, got %d.\n", (int)d, (int)d2);
+    }
+    else {
+        printf("    passed.\n");
+    }
+    size_t d3 = levenshtein_distance(a, a_len, b, b_len, buffer, d2);
+    assert(d2 == d3);
+    size_t d4 = levenshtein_distance(b, b_len, a, a_len, buffer, d2);
+    assert(d2 == d4);
+}
+
+static void
```

Comment:
Pleas remove these functions as we don't normally tests things this way (even with the macros).

I would recommend to either adapt this tests to sole NameError example's in `test_exceptions` or maybe expose them so they can be used in the  _testcapi module.